File: 2023/UICPC DIV1/UICPC Round #03/solutions/A/antimatterrain-zf.py
#!/usr/bin/python3

L = 19
MAXN = (1<<L)
INF = 1<<30

# i, x1, x2, y
NONE = (-1, 0, 0, INF)

# INPUT

D, S = map(int, input().split())

objs = []
xvals = set()
for i in range(D):
    x, y = map(int, input().split())
    xvals.add(x)
    objs.append((i, x, -INF, y))

for i in range(D, D+S):
    x1, x2, y = map(int, input().split())
    xvals.update((x1, x2))
    objs.append((i, x1, x2, y))

# COMPRESSION

xvals = sorted(xvals)
xmap = {xvals[i]:i for i in range(len(xvals))}
xmap[-INF] = -INF
mobjs = []
for o in objs: mobjs.append((o[0], xmap[o[1]], xmap[o[2]], o[3]))
objs = mobjs

# INITIALIZATION

rmq = [NONE]*(2*MAXN)
stacks = {i:[NONE] for i in range(MAXN)}
die = [0]*D
ykey = (lambda o : o[3])

# RMQ FUNCTIONS
def update(i, obj, lq = 0, hq = MAXN, node = 0):
    if lq+1 == hq:
        rmq[node] = obj
    else:
        left, mid, right = (node<<1)+1, (lq+hq)>>1, (node<<1)+2
        if i < mid: update(i, obj, lq, mid, left)
        else: update(i, obj, mid, hq, right)
        rmq[node] = min(rmq[left], rmq[right], key = ykey)

def query(lo, hi, lq = 0, hq = MAXN, node = 0):
    if hi <= lq or hq <= lo: return NONE
    if lo <= lq and hq <= hi: return rmq[node]

    mid = (lq+hq)>>1
    obj_l = query(lo, hi, lq, mid, (node<<1)+1)
    obj_r = query(lo, hi, mid, hq, (node<<1)+2)

    return min(obj_l, obj_r, key = ykey)

# PROCESSING FUNCTIONS
def addDrop(obj):
    stacks[obj[1]].append(obj)
    update(obj[1], obj)

def processSensor(obj):
    orig = query(obj[1], obj[2]+1)
    if orig == NONE: return

    hit = orig
    while hit[3] == orig[3]:
        die[hit[0]] = obj[3]
        stacks[hit[1]].pop()
        update(hit[1], stacks[hit[1]][-1])
        hit = query(obj[1], obj[2]+1)

# PROCESS
for obj in sorted(objs, key = ykey)[::-1]:
    if obj[2] == -INF: addDrop(obj)
    else: processSensor(obj)

print(*die, sep='\n')

File: 2023/UICPC DIV1/UICPC Round #03/solutions/A/antimatterrain-zf.cpp
#include <bits/stdc++.h>

using namespace std;

#define L 19
#define MAXN (1<<L)
#define INF (1<<30)

struct object {
  int i, x1, x2, y; // x2 == -INF indicates a point
  bool operator<(const object& rhs) const { return y < rhs.y; }
};

const object NONE = {-1, 0, 0, INF};

// x_to_i compresses the distinct x-values in the input
unordered_map<int, int> x_to_i;

// for x-value represented by compressed index i, st[i] is the stack of points
// with this x-value that have been swept over but not yet destroyed
stack<object> st[MAXN];

// rmq: given a range [lo, hi), will return the point with minimum y-value
// in the range among all points we have swept over but not yet destroyed
object rmq[2*MAXN];

// records the y-value when point i dies.
vector<int> die;

// update the object with compressed x-position i to "obj"
void update(int i, object obj, int lq = 0, int hq = MAXN, int node = 0) {
  if (lq+1 == hq) {
    rmq[node] = obj;
    return;
  }

  int mid = (lq+hq)>>1, left = (node<<1)+1, right = (node<<1)+2;

  if (i < mid) update(i, obj, lq, mid, left);
  else update(i, obj, mid, hq, right);

  rmq[node] = min(rmq[left], rmq[right]);
}

// find an object with minimum y-coordinate spanned by the query range
object query(int lo, int hi, int lq = 0, int hq = MAXN, int node = 0) {
  if (hi <= lq || hq <= lo) return NONE;
  if (lo <= lq && hq <= hi) return rmq[node];

  int mid = (lq+hq)>>1;
  object obj_l = query(lo, hi, lq, mid, (node<<1)+1);
  object obj_r = query(lo, hi, mid, hq, (node<<1)+2);

  return min(obj_l, obj_r);
}

void addPoint(object obj) {
  int i = x_to_i[obj.x1];
  st[i].push(obj);
  update(i, obj);
}

void processBarrier(object obj) {
  int il = x_to_i[obj.x1], ih = x_to_i[obj.x2]+1;

  object orig = query(il, ih);
  if (orig.i == NONE.i) return;

  // continue killing off points as long as there is one in this range
  // with the same x-value as the first point to hit the wall
  for (object hit = orig; hit.y == orig.y; hit = query(il, ih)) {
    // record this point hit a wall
    die[hit.i] = obj.y;

    int i = x_to_i[hit.x1];
    assert(st[i].size() > 0 && st[i].top().i != -1);
    st[i].pop();

    // update the rmq data for the next point with y-coordinate i
    update(i, st[i].top());
  }

}

int main() {
  int n, m;
  cin >> n >> m;

  vector<object> obj(n+m);
  die.assign(n, 0);

  for (int i = 0; i < n; ++i) {
    cin >> obj[i].x1 >> obj[i].y;
    obj[i].i = i;
    obj[i].x2 = -INF;
  }

  for (int i = n; i < m+n; ++i) cin >> obj[i].x1 >> obj[i].x2 >> obj[i].y;

  sort(obj.begin(), obj.end());

  // determine the distinct x-values (compression)
  set<int> xvals;
  for (auto o : obj) {
    xvals.insert(o.x1);
    if (o.x2 != -INF) xvals.insert(o.x2);
  }
  for (int x : xvals) x_to_i[x] = x_to_i.size();
  assert(xvals.size() <= MAXN);

  // initialize RMQ and stack
  fill(rmq, rmq+2*MAXN, NONE);
  for (auto& s : st) s.push(NONE);

  // finally do the sweep
  for (int i = obj.size()-1; i >= 0; --i) {
    if (obj[i].x2 == -INF) addPoint(obj[i]);
    else processBarrier(obj[i]);
  }

  for (int y : die) cout << y << endl;

  return 0;
}
